pkg/controllers/subnetset/subnetport_handler.go

```diff
@@ -0,0 +1,81 @@
+package subnetset
+
+import (
+	"context"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+)
+
+// SubnetPortHandler supports lazy-creation of Subnet, the first Subnet won't
+// be created until there is a SubnetPort attached to it.
+// - SubnetPort creation: get available Subnet for the SubnetPort, create new
+//   Subnet if necessary.
+// - SubnetPort deletion: if recycling Subnet is required, delete Subnets without
+// SubnetPort attached to it.
+
+type SubnetPortHandler struct {
+	Reconciler *SubnetSetReconciler
+}
+
+// Create allocates Subnet for SubnetPort from SubnetSet.
+func (h *SubnetPortHandler) Create(e event.CreateEvent, _ workqueue.RateLimitingInterface) {
+	subnetPort := e.Object.(*v1alpha1.SubnetPort)
+	if subnetPort.Spec.Subnet != "" {
+		// Two possible scenarios:
+		// - 1. User uses `.Spec.Subnet` directly instead of `.Spec.SubnetSet`.
+		// - 2. Subnet has been allocated and `.Spec.Subnet` is rendered by SubnetPortHandler.
+		return
+	}
+	subnetSet := &v1alpha1.SubnetSet{}
+	key := types.NamespacedName{
+		Namespace: subnetPort.GetNamespace(),
```

Comment:
For system VPC, the VPC CR will be created in one of system NS, but subnet port created in any of the system NS without a SubnetSet in spec should be matched to the same default subnet set created by vpc handler in the same NS as the system VPC CR.

While I don't think customers are allowed to create subnetset CR in system VPC, can we also confirm the behavior for non-default subnet set created in system NS? Should it be allowed for cross-NS sharing (suppose all of them are created in the same NS as system VPC). cc @lxiaopei

Reply:
Thanks for reminder, will confirm the bahaviour.

Reply:
Marking as resolved as this part of logic will be handled in Pod/SubnetPort controller, the subnetport_handler will be removed.